src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -1542,7 +1542,7 @@ trait Types
       case _                => lo <:< that && that <:< hi
     }
     private def emptyLowerBound = typeIsNothing(lo) || lo.isWildcard
-    private def emptyUpperBound = typeIsAny(hi) || hi.isWildcard
+    private def emptyUpperBound = typeIsAny(hi) || hi.eq(definitions.ObjectTpeJava) || hi.isWildcard
```

Comment:
should we actually move `hi.eq(definitions.ObjectTpeJava)` into `typeIsAny`?

Reply:
Added that in my squashed version

Reply:
Well, that didn't work -- compiler won't bootstrap. Didn't investigate.

Reply:
The (first?) bootstrap error:

```
src/library/scala/collection/concurrent/TrieMap.scala:285: error: type mismatch;
 found   : Any
 required: AnyRef
        ln.get(k).asInstanceOf[Option[AnyRef]].getOrElse(NO_SUCH_ELEMENT_SENTINEL)
                                                        ^
one error found
```

Amounts to:

```
$ tail sandbox/J.java sandbox/test.scala
==> sandbox/J.java <==
public class J {
   static final Object x = new Object();
}
==> sandbox/test.scala <==
class Test {
  def test(c: Option[AnyRef]): AnyRef = {
    c.getOrElse(J.x)
  }
}
```

I pushed a commit that pushed the check into `typeIsAny` except in the one place that doesn't make sense.
```